netebpfext/net_ebpf_ext.c

```diff
@@ -69,35 +63,138 @@ static net_ebpf_ext_wfp_callout_state_t _net_ebpf_ext_wfp_callout_state[] = {
-        L"L2 Callout",
-        L"L2 callout driver",
+        L"L2 Inbound",
+        L"L2 Inbound Callout for eBPF",
         FWP_ACTION_CALLOUT_TERMINATING,
     },
     {
-        &EBPF_HOOK_ALE_RESOURCE_ALLOC_CALLOUT,
+        &EBPF_HOOK_ALE_RESOURCE_ALLOC_V4_CALLOUT,
         &FWPM_LAYER_ALE_RESOURCE_ASSIGNMENT_V4,
         net_ebpf_ext_resource_allocation_classify,
         _net_ebpf_ext_filter_change_notify,
         _net_ebpf_ext_flow_delete,
-        L"Resource Allocation eBPF Callout",
-        L"Resource Allocation callout driver for eBPF",
+        L"Resource Allocation v4",
+        L"Resource Allocation v4 callout for eBPF",
         FWP_ACTION_CALLOUT_TERMINATING,
     },
     {
-        &EBPF_HOOK_ALE_RESOURCE_RELEASE_CALLOUT,
+        &EBPF_HOOK_ALE_RESOURCE_RELEASE_V4_CALLOUT,
         &FWPM_LAYER_ALE_RESOURCE_RELEASE_V4,
         net_ebpf_ext_resource_release_classify,
         _net_ebpf_ext_filter_change_notify,
         _net_ebpf_ext_flow_delete,
-        L"Resource Release eBPF Callout",
-        L"Resource Release callout driver for eBPF",
+        L"Resource Release v4",
+        L"Resource Release v4 callout for eBPF",
+        FWP_ACTION_CALLOUT_TERMINATING,
+    },
+    {
+        &EBPF_HOOK_ALE_RESOURCE_ALLOC_V6_CALLOUT,
+        &FWPM_LAYER_ALE_RESOURCE_ASSIGNMENT_V6,
+        net_ebpf_ext_resource_allocation_classify,
+        _net_ebpf_ext_filter_change_notify,
+        _net_ebpf_ext_flow_delete,
+        L"Resource Allocation v6",
+        L"Resource Allocation v6 callout for eBPF",
         FWP_ACTION_CALLOUT_TERMINATING,
     },
-};
+    {
+        &EBPF_HOOK_ALE_RESOURCE_RELEASE_V6_CALLOUT,
+        &FWPM_LAYER_ALE_RESOURCE_RELEASE_V6,
+        net_ebpf_ext_resource_release_classify,
+        _net_ebpf_ext_filter_change_notify,
+        _net_ebpf_ext_flow_delete,
+        L"Resource Release eBPF Callout v6",
+        L"Resource Release callout for eBPF",
+        FWP_ACTION_CALLOUT_TERMINATING,
+    }};
 
-// Callout globals
+// WFP globals
 static HANDLE _fwp_engine_handle;
 
+//
+// WFP component management related utility functions.
+//
+
+void
+net_ebpf_extension_delete_wfp_filters(uint32_t filter_count, _In_count_(filter_count) uint64_t* filter_ids)
+{
+    for (uint32_t index = 0; index <= filter_count; index++)
+        FwpmFilterDeleteById(_fwp_engine_handle, filter_ids[index]);
+}
+
+ebpf_result_t
+net_ebpf_extension_add_wfp_filters(
+    uint32_t filter_count,
+    _In_count_(filter_count) const net_ebpf_extension_wfp_filter_parameters_t* parameters,
+    uint32_t condition_count,
+    _In_opt_count_(condition_count) const FWPM_FILTER_CONDITION* conditions,
+    _In_ const void* raw_context,
+    _Out_writes_(filter_count) uint64_t* filter_ids)
+{
+    NTSTATUS status = STATUS_SUCCESS;
+    ebpf_result_t result = EBPF_SUCCESS;
+    BOOL is_in_transaction = FALSE;
+
+    status = FwpmTransactionBegin(_fwp_engine_handle, 0);
+    if (!NT_SUCCESS(status)) {
+        KdPrintEx(
```

Comment:
should this use EBPF_LOG_MESSAGE()?  Same question for line 167 etc

Reply:
netebpfext does not have a TraceLog provider yet. I will file an issue for this.

Reply:
https://github.com/microsoft/ebpf-for-windows/issues/805